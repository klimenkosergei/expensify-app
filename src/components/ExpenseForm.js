import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }
  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(prevState => ({ description }));
  };
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(prevState => ({ amount }));
    }
  };
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(prevState => ({ note }));
  };
  onDateChange = date => {
    if (date) {
      this.setState(prevState => ({ createdAt: date }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(prevState => ({ calendarFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(prevState => ({
        error: 'Please provide description and amount.'
      }));
    } else {
      // Clear error state
      this.setState(prevState => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount: parseFloat(this.state.amount) * 100, // Times 100 since our format is in pennies
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit} className='form'>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
            className='form__input'
          />
          <input
            type='text'
            placeholder='Amount'
            value={this.state.amount}
            onChange={this.onAmountChange}
            className='form__input'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            displayFormat={() => 'DD/MM/Y'}
            firstDayOfWeek={1}
            hideKeyboardShortcutsPanel={true}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder='Add a note to expense (optional)'
            value={this.state.note}
            onChange={this.onNoteChange}
            rows='4'
            className='form__textarea'
          ></textarea>
          <button type='submit' className='form__submit'>
            Submit
          </button>
        </form>
      </div>
    );
  }
}
