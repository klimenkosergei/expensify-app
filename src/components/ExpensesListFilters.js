import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';

import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

class ExpensesListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      calendarFocused: null
    };
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = calendarFocused => {
    this.setState(prevState => ({ calendarFocused }));
  };
  onTextChange = e => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = e => {
    if (e.target.value === 'date') {
      this.props.sortByDate();
    } else if (e.target.value === 'amount') {
      this.props.sortByAmount();
    }
  };
  render() {
    return (
      <div className='filters'>
        <div className='filters__wrapper'>
          <input
            type='text'
            value={this.props.filters.text}
            onChange={this.onTextChange}
            className='filters__text-filter'
            placeholder='Search expenses'
          />
          <select
            value={this.props.filters.sortBy}
            onChange={this.onSortChange}
            className='filters__sort-filter'
          >
            <option value='date'>Date</option>
            <option value='amount'>Amount</option>
          </select>
          <DateRangePicker
            startDate={this.props.filters.startDate}
            startDateId='startDateId'
            endDate={this.props.filters.endDate}
            endDateId='endDateId'
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            showClearDates={true}
            isOutsideRange={() => false}
            firstDayOfWeek={1}
            displayFormat={() => 'DD/MM/Y'}
            numberOfMonths={1}
            hideKeyboardShortcutsPanel={true}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: endDate => {
    dispatch(setEndDate(endDate));
  },
  setTextFilter: value => {
    dispatch(setTextFilter(value));
  },
  sortByDate: () => {
    dispatch(sortByDate());
  },
  sortByAmount: () => {
    dispatch(sortByAmount());
  }
});

const ConnectedExpensesListFilters = connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpensesListFilters);

export { ExpensesListFilters, ConnectedExpensesListFilters as default };
