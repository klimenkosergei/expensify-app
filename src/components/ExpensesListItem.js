import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpensesListItem = ({ id, description, amount, createdAt }) => (
  <div className='list__item'>
    <Link to={`/edit/${id}`} className='list__item-link'>
      <div className='list__item-column'>
        <h3 className='list__item-description'>{description}</h3>
        <span className='list__item-date'>
          {moment(createdAt).format('MMMM Do, Y')}
        </span>
      </div>
      <div className='list__item-column'>
        <span className='list__item-amount'>
          {numeral(amount / 100).format('$0,0.00')}
        </span>
      </div>
    </Link>
  </div>
);

export default ExpensesListItem;
