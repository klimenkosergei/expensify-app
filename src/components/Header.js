import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogOut } from '../actions/auth';

const Header = ({ logOut }) => (
  <header className='header'>
    <div className='header__wrapper'>
      <h1 className='header__title'>BudgetApp</h1>
      <NavLink
        to='/dashboard'
        className='header__link'
        activeClassName='is-active'
      >
        Dashboard
      </NavLink>
      <NavLink
        to='/create'
        className='header__link'
        activeClassName='is-active'
      >
        Create Expense
      </NavLink>
      <button className='header__btn' onClick={logOut}>
        Log Out
      </button>
    </div>
  </header>
);

const mapDispatchToProps = dispatch => ({
  logOut: () => {
    dispatch(startLogOut());
  }
});

const ConnectedHeader = connect(
  null,
  mapDispatchToProps
)(Header);

export { Header, ConnectedHeader as default };
