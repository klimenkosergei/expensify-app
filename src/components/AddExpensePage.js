import React from 'react';
import { connect } from 'react-redux';

import { startAddExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

class AddExpensePage extends React.Component {
  constructor(props) {
    super(props);
  }
  addExpense = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/projects/expensify');
  };
  render() {
    return (
      <div className='create'>
        <div className='create__wrapper'>
          <h2 className='create__title'>Create Expense</h2>
          <ExpenseForm onSubmit={this.addExpense} />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => {
    dispatch(startAddExpense(expense));
  }
});

const ConnectedAddExpensePage = connect(
  null,
  mapDispatchToProps
)(AddExpensePage); // Pass null for the mapStateToProps

export { AddExpensePage, ConnectedAddExpensePage as default };
