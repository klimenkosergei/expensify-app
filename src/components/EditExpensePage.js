import React from 'react';
import { connect } from 'react-redux';

import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
  }
  editExpense = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/projects/expensify');
  };
  removeExpense = () => {
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/projects/expensify');
  };
  render() {
    return (
      <div className='create'>
        <div className='create__wrapper'>
          <h2 className='create__title'>Edit Expense</h2>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.editExpense}
          />
          <button onClick={this.removeExpense} className='create__btn'>
            Delete
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  editExpense: (expenseId, expense) => {
    dispatch(startEditExpense(expenseId, expense));
  },
  removeExpense: expenseId => {
    dispatch(startRemoveExpense(expenseId));
  }
});

const ConnectedEditExpensePage = connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);

export { EditExpensePage, ConnectedEditExpensePage as default };
