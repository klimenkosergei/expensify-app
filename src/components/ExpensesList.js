import React from 'react';
import { connect } from 'react-redux';

import getVisibleExpenses from '../selectors/expenses';
import ExpensesListItem from './ExpensesListItem';

const ExpensesList = props => (
  <div className='list'>
    <div className='list__wrapper'>
      <div className='list__header'>
        <span className='list__header-item'>Expenses</span>
        <span className='list__header-item'>Amount</span>
      </div>
      {props.expenses.length !== 0 ? (
        props.expenses.map(expense => (
          <ExpensesListItem key={expense.id} {...expense} />
        ))
      ) : (
        <p className='list__item list__item--empty'>No expenses found</p>
      )}
    </div>
  </div>
);

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

const ConnectedExpensesList = connect(mapStateToProps)(ExpensesList);

export { ExpensesList, ConnectedExpensesList as default };
