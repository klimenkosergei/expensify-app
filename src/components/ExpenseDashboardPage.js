import React from 'react';
import ExpensesList from './ExpensesList';
import ExpensesSummary from './ExpensesSummary';
import ExpensesListFilters from './ExpensesListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpensesSummary />
    <ExpensesListFilters />
    <ExpensesList />
  </div>
);

export default ExpenseDashboardPage;
