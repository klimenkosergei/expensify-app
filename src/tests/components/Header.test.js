import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/Header';

let wrapper, logOut;

beforeEach(() => {
  logOut = jest.fn();
  wrapper = shallow(<Header logOut={logOut} />);
});

test('should render Header correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onClick for logOut', () => {
  wrapper.find('button').simulate('click');
  expect(logOut).toHaveBeenCalled();
});
