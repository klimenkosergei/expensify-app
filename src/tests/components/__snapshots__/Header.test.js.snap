// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header correctly 1`] = `
<header
  className="header"
>
  <div
    className="header__wrapper"
  >
    <h1
      className="header__title"
    >
      BudgetApp
    </h1>
    <NavLink
      activeClassName="is-active"
      className="header__link"
      to="/dashboard"
    >
      Dashboard
    </NavLink>
    <NavLink
      activeClassName="is-active"
      className="header__link"
      to="/create"
    >
      Create Expense
    </NavLink>
    <button
      className="header__btn"
      onClick={[MockFunction]}
    >
      Log Out
    </button>
  </div>
</header>
`;
